mount/upgrader_test.go

```diff
@@ -1,9 +1,238 @@
+				require.Error(t, err)
+			},
+			expectedContentFilePath: "../" + testdata.RootPathCarV1,
+		},
+
+		"transient file is copied from user's initial file": {
+			initial: "../" + testdata.RootPathCarV1,
+
+			createMnt: func(t *testing.T, key string, rootDir string) Mount {
+				return &FSMount{testdata.FS, testdata.FSPathCarV2} // purposely giving a different file here.
+			},
+
+			verify: func(t *testing.T, u *Upgrader, key string, rootDir string) {
+				_, err := os.Stat(u.TransientPath())
+				require.NoError(t, err)
+
+				// read the contents of the transient file.
+				tf, err := os.Open(u.TransientPath())
+				require.NoError(t, err)
+				defer tf.Close()
+				bz, err := ioutil.ReadAll(tf)
+				require.NoError(t, err)
+				require.NoError(t, tf.Close())
+
+				// read the contents of the initial file -> they should match.
+				f, err := os.Open("../" + testdata.RootPathCarV1)
+				require.NoError(t, err)
+				defer f.Close()
+				bz2, err := ioutil.ReadAll(f)
+				require.NoError(t, err)
+				require.NoError(t, f.Close())
+				require.EqualValues(t, bz, bz2)
+			},
+			expectedContentFilePath: "../" + testdata.RootPathCarV1,
+		},
+		"delete transient": {
+			setup: nil,
+			createMnt: func(t *testing.T, key string, rootDir string) Mount {
+				return &FSMount{testdata.FS, testdata.FSPathCarV2}
+			},
+			verify: func(t *testing.T, u *Upgrader, key string, rootDir string) {
+				ustat, err := u.Stat(context.TODO())
+				require.NoError(t, err)
+
+				fstat, err := os.Stat(u.TransientPath())
+				require.NoError(t, err)
+				require.EqualValues(t, fstat.Size(), ustat.Size)
+
+				err = u.DeleteTransient()
+				require.NoError(t, err)
+
+				_, err = os.Stat(u.TransientPath())
+				require.Error(t, err)
+
+				require.Empty(t, u.TransientPath())
+			},
+			expectedContentFilePath: "../" + testdata.RootPathCarV2,
+		},
+	}
+
+	ctx := context.Background()
+
+	for name, tc := range tcs {
+		tcc := tc
+		t.Run(name, func(t *testing.T) {
+			key := fmt.Sprintf("%d", rand.Uint64())
+			rootDir := t.TempDir()
+			if tcc.setup != nil {
+				tcc.setup(t, key, rootDir)
+			}
+
+			mnt := tcc.createMnt(t, key, rootDir)
+
+			u, err := Upgrade(mnt, rootDir, key, tcc.initial)
+			require.NoError(t, err)
+			require.NotNil(t, u)
+
+			// fetch and verify contents
+			rd, err := u.Fetch(ctx)
+			require.NoError(t, err)
+			require.NotNil(t, rd)
+
+			bz, err := ioutil.ReadAll(rd)
+			require.NoError(t, err)
+			require.NotEmpty(t, bz)
+			require.NoError(t, rd.Close())
+
+			f, err := os.Open(tcc.expectedContentFilePath)
+			require.NoError(t, err)
+			bz2, err := ioutil.ReadAll(f)
+			require.NoError(t, err)
+			require.NoError(t, f.Close())
+			require.EqualValues(t, bz2, bz)
+
+			tcc.verify(t, u, key, rootDir)
+		})
+	}
+}
+
+type countingMount struct {
+	mu sync.Mutex
+	n  int
+
+	Mount
+}
+
+func (c *countingMount) Fetch(ctx context.Context) (Reader, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.n++
+
+	return c.Mount.Fetch(ctx)
+}
```

Comment:
This is incorrect, as the exclusive lock _also_ guards the mount fetch (thus we're not testing the deduplication of the mount!). This counter is better handled with an atomic int. Will change in a subsequent PR.